docs/source/rfc-configuration-overhaul.md

```diff
@@ -0,0 +1,721 @@
+
+> **Sumeragi: default `null` values**
+>
+> A special note about sumeragi fields with `null` as default: only the `trusted_peers` field out of the three can be initialized via a provided file or an environment variable.
+>
+> The other two fields, namely `key_pair` and `peer_id`, go through a process of finalization where their values are derived from the corresponding ones in the uppermost Iroha config (using its `public_key` and `private_key` fields) or the Torii config (via its `p2p_addr`). This ensures that these linked fields stay in sync, and prevents the programmer error when different values are provided to these field pairs. Providing either `sumeragi.key_pair` or `sumeragi.peer_id` by hand will result in an error, as it should never be done directly.
+
+As we can see, if the user tries to initialize `key_pair` and `peer_id` via the config file or env variables, they'll get an error. We are creating a terrible user experience by presenting the user with an option to configure something they should not be configuring.
+
+We should avoid exposing fields like this to the user.
+
+### Naming Issues
+
+The configuration parameters naming presents a few challenges.
+
+#### Overcomplicated ENV Variables
+
+ENV variables for all configuration parameters are named according to the internal module names. These technical designations might resonate with developers but can be confusing or alienating for users not deeply immersed in the project's inner mechanisms. Many FOSS projects gravitate towards generic, descriptive names to ensure broader comprehension and enhanced usability.
+
+#### Inconsistent Naming of ENV Variables
+
+The output from the `iroha --help` command displays a mix of naming conventions:
+
+```
+Iroha 2 is configured via environment variables:
+    IROHA2_CONFIG_PATH is the location of your `config.json` or `config.json5`
+    IROHA2_GENESIS_PATH is the location of your `genesis.json` or `genesis.json5`
+...
+    IROHA_TORII: Torii (gateway) endpoint configuration
+    IROHA_SUMERAGI: Sumeragi (emperor) consensus configuration
+    IROHA_KURA: Kura (storage). Configuration of block storage
+    IROHA_BLOCK_SYNC: Block synchronisation configuration
+...
+```
+
+The inconsistency between the `IROHA2` and `IROHA` prefixes can be confusing.
+
+#### Inconsistent Naming of Configuration Parameters
+
+Current configuration demonstrates some inconsistencies in parameter naming, leading to potential user confusion. Specifically:
+
+- In the `torii` section:
+  - We have terms like `p2p_addr`, `api_url`, and `telemetry_url`. While "addr" and "url" may convey similar intents, the interchangeability of these terms is non-intuitive.
+  - Using prefixes instead of suffixes, like transitioning from `p2p_addr` to `addr_p2p`, progresses from a general term to a specific one, simplifying interpretation.
```

Comment:
Furthermore, I'd argue that `p2p` is the more general term here if interpreted in context of iroha. I feel like parameters are better off being grouped by subsystem rather than by name of knob they tweak.

Reply:
I think we can get rid of the internal codename as well, so something along the lines of 
```toml
telemetry.addr = "localhost:8081"
```

Reply:
Addressed in the document.